<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.1" />
<title>phc.easy.omics.options.genomic_short_variant API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>.flex{display:flex !important}body{line-height:1.5em;font-family:"Lato",Helvetica,Roboto,Arial,"Lucida Grande",sans-serif}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:90%}ul li code a{font-size:80%}a{color:#00539a;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#f7944d}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
<link rel="shortcut icon" href="./favicon.ico">
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>phc.easy.omics.options.genomic_short_variant</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from enum import Enum
from typing import List, Optional, Union

from phc.easy.omics.options.coding_effect import CodingEffect
from phc.easy.omics.options.chromosome import Chromosome
from phc.easy.omics.options.clinvar_significance import ClinVarSignificance
from phc.easy.omics.options.clinvar_review import ClinVarReview
from phc.easy.omics.options.gene_class import GeneClass
from phc.easy.omics.options.zygosity import Zygosity
from phc.easy.omics.options.common import GenomicVariantInclude
from phc.easy.abstract.paging_api_item import PagingApiOptions
from pydantic import Field, constr

RS_ID = r&#34;^rs(\d+)$&#34;
NUM_DECIMAL_RANGE = constr(regex=r&#34;^\d+(\.\d+)?\-\d+(\.\d+)?$&#34;)
NUM_RANGE = constr(regex=r&#34;^(\d+\-\d+|\d+)$&#34;)

MAPPINGS = {
    &#34;variant_set_ids&#34;: &#34;variantSetIds&#34;,
    &#34;rs_id&#34;: &#34;rsid&#34;,
    &#34;clinvar_allele_id&#34;: &#34;clinvarAlleleId&#34;,
    &#34;clinvar_disease&#34;: &#34;clinvarDisease&#34;,
    &#34;clinvar_review&#34;: &#34;clinvarReview&#34;,
    &#34;clinvar_significance&#34;: &#34;clinvarSignificance&#34;,
    &#34;cosmic_id&#34;: &#34;cosmicId&#34;,
    &#34;cosmic_status&#34;: &#34;cosmicStatus&#34;,
    &#34;cosmic_histology&#34;: &#34;cosmicHistology&#34;,
    &#34;cosmic_tumor_site&#34;: &#34;cosmicTumorSite&#34;,
    &#34;coding_effect&#34;: &#34;group&#34;,
    &#34;variant_class&#34;: &#34;class&#34;,
    &#34;transcript_id&#34;: &#34;transcriptId&#34;,
    &#34;gene_class&#34;: &#34;biotype&#34;,
    &#34;protein_changes&#34;: &#34;aminoAcidChange&#34;,
    &#34;sequence_type&#34;: &#34;sequenceType&#34;,
    # Used cosmic_min_count instead to match PHC interface
    # &#34;cosmic_sample_count&#34;: &#34;cosmicSampleCount&#34;,
    &#34;min_allele_frequency&#34;: &#34;minAlleleFrequency&#34;,
    &#34;max_allele_frequency&#34;: &#34;maxAlleleFrequency&#34;,
    &#34;pop_allele_frequency&#34;: &#34;popAlleleFrequency&#34;,
    &#34;exac_allele_frequency&#34;: &#34;exacAlleleFrequency&#34;,
    &#34;exac_homozygous&#34;: &#34;exacHomozygous&#34;,
    &#34;dbnsfp_damaging_count&#34;: &#34;dbnsfpDamagingCount&#34;,
    &#34;dbnsfp_damaging_predictor&#34;: &#34;dbnsfpDamagingPredictor&#34;,
    &#34;dbnsfp_damaging_vote&#34;: &#34;dbnsfpDamagingVote&#34;,
    &#34;dbnsfp_fathmm_rankscore&#34;: &#34;dbnsfpFathmmRankscore&#34;,
    &#34;dbnsfp_fathmm_pred&#34;: &#34;dbnsfpFathmmPred&#34;,
    &#34;dbnsfp_mean_rankscore&#34;: &#34;dbnsfpMeanRankscore&#34;,
    &#34;dbnsfp_mean_rankscore_predictor&#34;: &#34;dbnsfpMeanRankscorePredictor&#34;,
    &#34;dbnsfp_mutationtaster_rankscore&#34;: &#34;dbnsfpMutationtasterRankscore&#34;,
    &#34;dbnsfp_mutationtaster_pred&#34;: &#34;dbnsfpMutationtasterPred&#34;,
    &#34;dbnsfp_sift_rankscore&#34;: &#34;dbnsfpSiftRankscore&#34;,
    &#34;dbnsfp_sift_pred&#34;: &#34;dbnsfpSiftPred&#34;,
    &#34;variant_allele_frequency&#34;: &#34;variantAlleleFrequency&#34;,
    &#34;read_depth&#34;: &#34;readDepth&#34;,
    &#34;alt_read_depth&#34;: &#34;altReadDepth&#34;,
    &#34;ref_read_depth&#34;: &#34;refReadDepth&#34;,
    &#34;variant_filter&#34;: &#34;filter&#34;,
    &#34;in_ckb&#34;: &#34;drugAssociations&#34;,
}


class GenomicShortVariantOptions(PagingApiOptions):
    &#34;&#34;&#34;Options to pass to `/v1/genomics/variants`

    See https://docs.us.lifeomic.com/api/#query-short-variant-data
    &#34;&#34;&#34;

    # TODO: Add remaining options from Omics Explorer in PHC
    # - Variant Quality
    # - Variant Allele Freq
    # - Combined In Silico Prediction
    # - Individual In Silico Predictors

    variant_set_ids: List[str] = Field(..., min_items=1)
    include: List[GenomicVariantInclude] = [&#34;vcf&#34;]
    gene: List[str] = []
    rs_id: List[constr(regex=RS_ID)] = []
    chromosome: List[Chromosome] = []
    clinvar_allele_id: List[str] = []
    clinvar_disease: List[str] = []
    clinvar_review: List[ClinVarReview] = []
    clinvar_significance: List[ClinVarSignificance] = []
    cosmic_id: List[str] = []
    cosmic_status: List[str] = []
    cosmic_histology: List[str] = []
    cosmic_tumor_site: List[str] = []
    variant_class: List[str] = []  # Renamed from &#39;class&#39;
    coding_effect: List[CodingEffect] = []
    impact: List[str] = []
    transcript_id: List[str] = []
    gene_class: List[GeneClass] = []
    protein_changes: List[str] = []
    sequence_type: List[str] = []
    position: List[Union[int, NUM_RANGE]] = []
    cosmic_min_count: Optional[int] = None
    min_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    max_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    pop_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    exac_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    exac_homozygous: List[str] = []
    dbnsfp_damaging_count: List[str] = []
    dbnsfp_damaging_predictor: List[str] = []
    dbnsfp_damaging_vote: List[str] = []
    dbnsfp_fathmm_rankscore: List[str] = []
    dbnsfp_fathmm_pred: List[str] = []
    dbnsfp_mean_rankscore: List[str] = []
    dbnsfp_mean_rankscore_predictor: List[str] = []
    dbnsfp_mutationtaster_rankscore: List[str] = []
    dbnsfp_mutationtaster_pred: List[str] = []
    dbnsfp_sift_rankscore: List[str] = []
    dbnsfp_sift_pred: List[str] = []
    zygosity: List[Zygosity] = []
    genotype: List[str] = []
    variant_allele_frequency: List[str] = []
    quality: List[str] = []
    read_depth: List[str] = []
    alt_read_depth: List[str] = []
    ref_read_depth: List[str] = []
    variant_filter: List[str] = []
    in_ckb: Optional[bool]

    @staticmethod
    def transform(key, value):
        if isinstance(value, list):
            value = &#34;,&#34;.join(
                [elem if isinstance(elem, str) else str(elem) for elem in value]
            )
        elif isinstance(value, bool):
            value = &#34;true&#34; if value else None

        if key == &#34;cosmic_min_count&#34; and value is not None:
            return (&#34;cosmicSampleCount&#34;, f&#34;{value}:gte&#34;)

        return (MAPPINGS.get(key, key), value)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue"><code class="flex name class">
<span>class <span class="ident">NUM_DECIMAL_RANGE</span></span>
<span>(</span><span>...)</span>
</code></dt>
<dd>
<div class="desc"><p>str(object='') -&gt; str
str(bytes_or_buffer[, encoding[, errors]]) -&gt; str</p>
<p>Create a new string object from the given object. If encoding or
errors is specified, then the object must expose a data buffer
that will be decoded using the given encoding and error handler.
Otherwise, returns the result of object.<strong>str</strong>() (if defined)
or repr(object).
encoding defaults to sys.getdefaultencoding().
errors defaults to 'strict'.</p></div>
<h3>Ancestors</h3>
<ul class="hlist">
<li>pydantic.types.ConstrainedStr</li>
<li>builtins.str</li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.curtail_length"><code class="name">var <span class="ident">curtail_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.max_length"><code class="name">var <span class="ident">max_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.min_length"><code class="name">var <span class="ident">min_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.regex"><code class="name">var <span class="ident">regex</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.strict"><code class="name">var <span class="ident">strict</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.strip_whitespace"><code class="name">var <span class="ident">strip_whitespace</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue"><code class="flex name class">
<span>class <span class="ident">NUM_RANGE</span></span>
<span>(</span><span>...)</span>
</code></dt>
<dd>
<div class="desc"><p>str(object='') -&gt; str
str(bytes_or_buffer[, encoding[, errors]]) -&gt; str</p>
<p>Create a new string object from the given object. If encoding or
errors is specified, then the object must expose a data buffer
that will be decoded using the given encoding and error handler.
Otherwise, returns the result of object.<strong>str</strong>() (if defined)
or repr(object).
encoding defaults to sys.getdefaultencoding().
errors defaults to 'strict'.</p></div>
<h3>Ancestors</h3>
<ul class="hlist">
<li>pydantic.types.ConstrainedStr</li>
<li>builtins.str</li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.curtail_length"><code class="name">var <span class="ident">curtail_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.max_length"><code class="name">var <span class="ident">max_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.min_length"><code class="name">var <span class="ident">min_length</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.regex"><code class="name">var <span class="ident">regex</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.strict"><code class="name">var <span class="ident">strict</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue.strip_whitespace"><code class="name">var <span class="ident">strip_whitespace</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions"><code class="flex name class">
<span>class <span class="ident">GenomicShortVariantOptions</span></span>
<span>(</span><span>**data: Any)</span>
</code></dt>
<dd>
<div class="desc"><p>Options to pass to <code>/v1/genomics/variants</code></p>
<p>See <a href="https://docs.us.lifeomic.com/api/#query-short-variant-data">https://docs.us.lifeomic.com/api/#query-short-variant-data</a></p>
<p>Create a new model by parsing and validating input data from keyword arguments.</p>
<p>Raises ValidationError if the input data cannot be parsed to form a valid model.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class GenomicShortVariantOptions(PagingApiOptions):
    &#34;&#34;&#34;Options to pass to `/v1/genomics/variants`

    See https://docs.us.lifeomic.com/api/#query-short-variant-data
    &#34;&#34;&#34;

    # TODO: Add remaining options from Omics Explorer in PHC
    # - Variant Quality
    # - Variant Allele Freq
    # - Combined In Silico Prediction
    # - Individual In Silico Predictors

    variant_set_ids: List[str] = Field(..., min_items=1)
    include: List[GenomicVariantInclude] = [&#34;vcf&#34;]
    gene: List[str] = []
    rs_id: List[constr(regex=RS_ID)] = []
    chromosome: List[Chromosome] = []
    clinvar_allele_id: List[str] = []
    clinvar_disease: List[str] = []
    clinvar_review: List[ClinVarReview] = []
    clinvar_significance: List[ClinVarSignificance] = []
    cosmic_id: List[str] = []
    cosmic_status: List[str] = []
    cosmic_histology: List[str] = []
    cosmic_tumor_site: List[str] = []
    variant_class: List[str] = []  # Renamed from &#39;class&#39;
    coding_effect: List[CodingEffect] = []
    impact: List[str] = []
    transcript_id: List[str] = []
    gene_class: List[GeneClass] = []
    protein_changes: List[str] = []
    sequence_type: List[str] = []
    position: List[Union[int, NUM_RANGE]] = []
    cosmic_min_count: Optional[int] = None
    min_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    max_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    pop_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    exac_allele_frequency: Optional[NUM_DECIMAL_RANGE] = None
    exac_homozygous: List[str] = []
    dbnsfp_damaging_count: List[str] = []
    dbnsfp_damaging_predictor: List[str] = []
    dbnsfp_damaging_vote: List[str] = []
    dbnsfp_fathmm_rankscore: List[str] = []
    dbnsfp_fathmm_pred: List[str] = []
    dbnsfp_mean_rankscore: List[str] = []
    dbnsfp_mean_rankscore_predictor: List[str] = []
    dbnsfp_mutationtaster_rankscore: List[str] = []
    dbnsfp_mutationtaster_pred: List[str] = []
    dbnsfp_sift_rankscore: List[str] = []
    dbnsfp_sift_pred: List[str] = []
    zygosity: List[Zygosity] = []
    genotype: List[str] = []
    variant_allele_frequency: List[str] = []
    quality: List[str] = []
    read_depth: List[str] = []
    alt_read_depth: List[str] = []
    ref_read_depth: List[str] = []
    variant_filter: List[str] = []
    in_ckb: Optional[bool]

    @staticmethod
    def transform(key, value):
        if isinstance(value, list):
            value = &#34;,&#34;.join(
                [elem if isinstance(elem, str) else str(elem) for elem in value]
            )
        elif isinstance(value, bool):
            value = &#34;true&#34; if value else None

        if key == &#34;cosmic_min_count&#34; and value is not None:
            return (&#34;cosmicSampleCount&#34;, f&#34;{value}:gte&#34;)

        return (MAPPINGS.get(key, key), value)</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="phc.easy.abstract.paging_api_item.PagingApiOptions" href="../../abstract/paging_api_item.html#phc.easy.abstract.paging_api_item.PagingApiOptions">PagingApiOptions</a></li>
<li>pydantic.main.BaseModel</li>
<li>pydantic.utils.Representation</li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.alt_read_depth"><code class="name">var <span class="ident">alt_read_depth</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.chromosome"><code class="name">var <span class="ident">chromosome</span> : List[<a title="phc.easy.omics.options.chromosome.Chromosome" href="chromosome.html#phc.easy.omics.options.chromosome.Chromosome">Chromosome</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.clinvar_allele_id"><code class="name">var <span class="ident">clinvar_allele_id</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.clinvar_disease"><code class="name">var <span class="ident">clinvar_disease</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.clinvar_review"><code class="name">var <span class="ident">clinvar_review</span> : List[<a title="phc.easy.omics.options.clinvar_review.ClinVarReview" href="clinvar_review.html#phc.easy.omics.options.clinvar_review.ClinVarReview">ClinVarReview</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.clinvar_significance"><code class="name">var <span class="ident">clinvar_significance</span> : List[<a title="phc.easy.omics.options.clinvar_significance.ClinVarSignificance" href="clinvar_significance.html#phc.easy.omics.options.clinvar_significance.ClinVarSignificance">ClinVarSignificance</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.coding_effect"><code class="name">var <span class="ident">coding_effect</span> : List[<a title="phc.easy.omics.options.coding_effect.CodingEffect" href="coding_effect.html#phc.easy.omics.options.coding_effect.CodingEffect">CodingEffect</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.cosmic_histology"><code class="name">var <span class="ident">cosmic_histology</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.cosmic_id"><code class="name">var <span class="ident">cosmic_id</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.cosmic_min_count"><code class="name">var <span class="ident">cosmic_min_count</span> : Union[int, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.cosmic_status"><code class="name">var <span class="ident">cosmic_status</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.cosmic_tumor_site"><code class="name">var <span class="ident">cosmic_tumor_site</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_damaging_count"><code class="name">var <span class="ident">dbnsfp_damaging_count</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_damaging_predictor"><code class="name">var <span class="ident">dbnsfp_damaging_predictor</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_damaging_vote"><code class="name">var <span class="ident">dbnsfp_damaging_vote</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_fathmm_pred"><code class="name">var <span class="ident">dbnsfp_fathmm_pred</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_fathmm_rankscore"><code class="name">var <span class="ident">dbnsfp_fathmm_rankscore</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_mean_rankscore"><code class="name">var <span class="ident">dbnsfp_mean_rankscore</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_mean_rankscore_predictor"><code class="name">var <span class="ident">dbnsfp_mean_rankscore_predictor</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_mutationtaster_pred"><code class="name">var <span class="ident">dbnsfp_mutationtaster_pred</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_mutationtaster_rankscore"><code class="name">var <span class="ident">dbnsfp_mutationtaster_rankscore</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_sift_pred"><code class="name">var <span class="ident">dbnsfp_sift_pred</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dbnsfp_sift_rankscore"><code class="name">var <span class="ident">dbnsfp_sift_rankscore</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.exac_allele_frequency"><code class="name">var <span class="ident">exac_allele_frequency</span> : Union[<a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.exac_homozygous"><code class="name">var <span class="ident">exac_homozygous</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.gene"><code class="name">var <span class="ident">gene</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.gene_class"><code class="name">var <span class="ident">gene_class</span> : List[<a title="phc.easy.omics.options.gene_class.GeneClass" href="gene_class.html#phc.easy.omics.options.gene_class.GeneClass">GeneClass</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.genotype"><code class="name">var <span class="ident">genotype</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.impact"><code class="name">var <span class="ident">impact</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.in_ckb"><code class="name">var <span class="ident">in_ckb</span> : Union[bool, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.include"><code class="name">var <span class="ident">include</span> : List[<a title="phc.easy.omics.options.common.GenomicVariantInclude" href="common.html#phc.easy.omics.options.common.GenomicVariantInclude">GenomicVariantInclude</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.max_allele_frequency"><code class="name">var <span class="ident">max_allele_frequency</span> : Union[<a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.min_allele_frequency"><code class="name">var <span class="ident">min_allele_frequency</span> : Union[<a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.pop_allele_frequency"><code class="name">var <span class="ident">pop_allele_frequency</span> : Union[<a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>, NoneType]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.position"><code class="name">var <span class="ident">position</span> : List[Union[int, <a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>]]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.protein_changes"><code class="name">var <span class="ident">protein_changes</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.quality"><code class="name">var <span class="ident">quality</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.read_depth"><code class="name">var <span class="ident">read_depth</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.ref_read_depth"><code class="name">var <span class="ident">ref_read_depth</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.rs_id"><code class="name">var <span class="ident">rs_id</span> : List[<a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.sequence_type"><code class="name">var <span class="ident">sequence_type</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.transcript_id"><code class="name">var <span class="ident">transcript_id</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.variant_allele_frequency"><code class="name">var <span class="ident">variant_allele_frequency</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.variant_class"><code class="name">var <span class="ident">variant_class</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.variant_filter"><code class="name">var <span class="ident">variant_filter</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.variant_set_ids"><code class="name">var <span class="ident">variant_set_ids</span> : List[str]</code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.zygosity"><code class="name">var <span class="ident">zygosity</span> : List[<a title="phc.easy.omics.options.zygosity.Zygosity" href="zygosity.html#phc.easy.omics.options.zygosity.Zygosity">Zygosity</a>]</code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Static methods</h3>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.transform"><code class="name flex">
<span>def <span class="ident">transform</span></span>(<span>key, value)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@staticmethod
def transform(key, value):
    if isinstance(value, list):
        value = &#34;,&#34;.join(
            [elem if isinstance(elem, str) else str(elem) for elem in value]
        )
    elif isinstance(value, bool):
        value = &#34;true&#34; if value else None

    if key == &#34;cosmic_min_count&#34; and value is not None:
        return (&#34;cosmicSampleCount&#34;, f&#34;{value}:gte&#34;)

    return (MAPPINGS.get(key, key), value)</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dict"><code class="name flex">
<span>def <span class="ident">dict</span></span>(<span>self)</span>
</code></dt>
<dd>
<p class="inheritance">
<em>Inherited from:</em>
<code><a title="phc.easy.abstract.paging_api_item.PagingApiOptions" href="../../abstract/paging_api_item.html#phc.easy.abstract.paging_api_item.PagingApiOptions">PagingApiOptions</a></code>.<code><a title="phc.easy.abstract.paging_api_item.PagingApiOptions.dict" href="../../abstract/paging_api_item.html#phc.easy.abstract.paging_api_item.PagingApiOptions.dict">dict</a></code>
</p>
<div class="desc inherited"><p>Generate a dictionary representation of the model, optionally specifying which fields to include or exclude.</p></div>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<header>
<a class="homelink" rel="home" title="PHC Home" href="https://lifeomic.github.io/phc-sdk-py/">
<img src="./phc.png" alt="">
</a>
</header>
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="phc.easy.omics.options" href="index.html">phc.easy.omics.options</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a></code></h4>
</li>
<li>
<h4><code><a title="phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue" href="#phc.easy.omics.options.genomic_short_variant.ConstrainedStrValue">ConstrainedStrValue</a></code></h4>
</li>
<li>
<h4><code><a title="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions" href="#phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions">GenomicShortVariantOptions</a></code></h4>
<ul class="">
<li><code><a title="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dict" href="../../abstract/paging_api_item.html#phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.dict">dict</a></code></li>
<li><code><a title="phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.transform" href="#phc.easy.omics.options.genomic_short_variant.GenomicShortVariantOptions.transform">transform</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.1</a>.</p>
</footer>
</body>
</html>